hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/SamlEngineModule.java

```diff
@@ -307,22 +302,24 @@ private MetadataHealthCheck getVerifyMetadataHealthCheck(
     }
 
     @Provides
+    @Named("VERIFY_METADATA_REFRESH_TASK")
     @Singleton
-    private IdpSingleSignOnServiceHelper getIdpSingleSignOnServiceHelper(@Named("VerifyMetadataResolver") MetadataResolver metadataResolver) {
-        return new IdpSingleSignOnServiceHelper(metadataResolver);
+    private Task registerMetadataRefreshTask(Environment environment, @Named(VERIFY_METADATA_RESOLVER) MetadataResolver metadataResolver) {
+        Task task = new Task("metadata-refresh") {
+            @Override
+            public void execute(ImmutableMultimap<String, String> parameters, PrintWriter output) throws Exception {
+                ((AbstractReloadingMetadataResolver) metadataResolver).refresh();
```

Comment:
Is it possible to avoid this cast by changing the parameter's type? Feels a bit weird to have a method that takes a `MetadataResolver` and only works properly if you pass the right subtype in.

Reply:
Part of the problem is that our factories for `MetadataResolver` only return a `MetadataResolver` rather than the type of what actually gets build `JerseyClientMetadataResolver` so we have to do this casting here. We'd have to change saml-lib to fix it.

At some point soon I'd like to move this binding behaviour into the `MetadataResolverBundle` where the refresh task will be added automatically without guice.

Reply:
Ok, that's fair enough then.